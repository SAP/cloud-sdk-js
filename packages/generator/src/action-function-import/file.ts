import { flat } from '@sap-cloud-sdk/util';
import {
  FunctionDeclarationStructure,
  InterfaceDeclarationStructure,
  SourceFileStructure,
  StructureKind
} from 'ts-morph';
import { VdmFunctionImport, VdmServiceMetadata } from '../vdm-types';
import { parametersInterface } from './parameters-interface';
import { exportStatement } from './export-statement';
import { actionImportFunction } from './action';
import { importDeclarationsAction, importDeclarationsFunction } from './import';
import { functionImportFunction } from './function';

/**
 * @internal
 */
export function actionImportSourceFile(
  service: VdmServiceMetadata
): SourceFileStructure {
  if (!service.actionImports) {
    throw new Error(
      'Tried to create action import source files without actions in service metadata.'
    );
  }
  return {
    kind: StructureKind.SourceFile,
    statements: [
      ...importDeclarationsFunction(service),
      ...flat(
        service.actionImports.map(action =>
          actionImportStatements(action, service)
        )
      ),
      exportStatement(service.actionImports, 'actionImports')
    ]
  };
}

function actionImportStatements(
  actionImport: VdmFunctionImport,
  service: VdmServiceMetadata
): [InterfaceDeclarationStructure, FunctionDeclarationStructure] {
  return [
    parametersInterface(actionImport),
    actionImportFunction(actionImport, service)
  ];
}
/**
 * @internal
 */
export function functionImportSourceFile(
  service: VdmServiceMetadata
): SourceFileStructure {
  return {
    kind: StructureKind.SourceFile,
    statements: [
      ...importDeclarationsAction(service),
      ...flat(
        service.functionImports.map(fi => functionImportStatements(fi, service))
      ),
      exportStatement(service.functionImports, 'functionImports')
    ]
  };
}

function functionImportStatements(
  functionImport: VdmFunctionImport,
  service: VdmServiceMetadata
): [InterfaceDeclarationStructure, FunctionDeclarationStructure] {
  return [
    parametersInterface(functionImport),
    functionImportFunction(functionImport, service)
  ];
}
