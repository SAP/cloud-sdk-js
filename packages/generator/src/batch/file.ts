import { SourceFileStructure, StructureKind } from 'ts-morph';
import { VdmServiceMetadata } from '../vdm-types';
import { batchFunction, changesetFunction } from './function';
import { importBatchDeclarations } from './imports';
import { readRequestType, writeRequestType } from './type';

/**
 * @internal
 */
export function batchSourceFile(
  service: VdmServiceMetadata
): SourceFileStructure {
  return {
    kind: StructureKind.SourceFile,
    statements: [
      ...importBatchDeclarations(service),
      batchFunction(service),
      changesetFunction(service),
      `export const default${service.className}Path = '${service.servicePath}';`,
      readRequestType(service),
      writeRequestType(service)
    ]
  };
}
