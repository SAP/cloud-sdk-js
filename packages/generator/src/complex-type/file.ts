import { SourceFileStructure, StructureKind } from 'ts-morph';
import { ODataVersion } from '@sap-cloud-sdk/util';
import { VdmComplexType } from '../vdm-types';
import { fieldTypeClass } from './field-type-class';
import { importDeclarations } from './imports';
import { complexTypeInterface } from './interface';
import { complexTypeNamespace } from './namespace';

/**
 * @internal
 */
export function complexTypeSourceFile(
  complexType: VdmComplexType,
  oDataVersion: ODataVersion
): SourceFileStructure {
  return {
    kind: StructureKind.SourceFile,
    statements: [
      ...importDeclarations(complexType, oDataVersion),
      complexTypeInterface(complexType),
      fieldTypeClass(complexType),
      complexTypeNamespace(complexType)
    ]
  };
}
