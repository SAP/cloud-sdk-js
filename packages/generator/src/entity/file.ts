import { SourceFileStructure, StructureKind } from 'ts-morph';
import { parametersInterface } from '../operations';
import { VdmEntity, VdmServiceMetadata } from '../vdm-types';
import { entityClass } from './class';
import { entityImportDeclarations, otherEntityImports } from './imports';
import { entityTypeInterface } from './interface';

/**
 * @internal
 */
export function entitySourceFile(
  entity: VdmEntity,
  service: VdmServiceMetadata
): SourceFileStructure {
  return {
    kind: StructureKind.SourceFile,
    statements: [
      ...entityImportDeclarations(entity, service, service.oDataVersion),
      ...otherEntityImports(entity, service),
      entityClass(entity, service),
      entityTypeInterface(entity, service),
      ...[...entity.functions, ...entity.actions].map(operation =>
        parametersInterface(operation)
      )
    ]
  };
}
