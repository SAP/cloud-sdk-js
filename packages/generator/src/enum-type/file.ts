import { SourceFileStructure, StructureKind } from 'ts-morph';
import { VdmEnumType } from '../vdm-types';
import { enumTypeClass } from './enum';

/**
 * @internal
 */
export function enumTypeSourceFile(enumType: VdmEnumType): SourceFileStructure {
  return {
    kind: StructureKind.SourceFile,
    statements: [enumTypeClass(enumType)]
  };
}
