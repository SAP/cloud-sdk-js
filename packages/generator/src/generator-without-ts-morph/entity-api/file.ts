import { unixEOL } from '@sap-cloud-sdk/util';
import {
  serializeImports,
  Import
} from '@sap-cloud-sdk/generator-common/internal';
import { VdmEntity, VdmServiceMetadata } from '../../vdm-types';
import {
  navPropertyFieldTypeImportNames,
  propertyFieldTypeImportNames,
  propertyTypeImportNames
} from '../../imports';
import { odataImport, complexTypeImports, enumTypeImports } from './imports';
import { classContent } from './class';

/**
 * @internal
 */
export function entityApiFile(
  entity: VdmEntity,
  service: VdmServiceMetadata
): string {
  const imports = serializeImports(getImports(entity, service));
  const content = classContent(entity, service);
  return [imports, content].join(unixEOL);
}

// todo Use this function instead the one above, when the todo of "de-serializers.ts" is solved, so that unnecessary external dependencies are not imported.
// export function file(entity: VdmEntity, service: VdmServiceMetadata): string {
//   const imports = serializeImports([...getImports(entity, service), ...externalImports(entity.properties)]);
//   const content = classContent(entity, service);
//   return [
//     imports,
//     content
//   ].join(unixEOL);
// }

function getImports(entity: VdmEntity, service: VdmServiceMetadata): Import[] {
  return [
    {
      names: [`${entity.className}`],
      moduleIdentifier: `./${entity.className}`,
      typeOnly: false
    },
    {
      names: [`${entity.className}RequestBuilder`],
      moduleIdentifier: `./${entity.className}RequestBuilder`,
      typeOnly: false
    },
    ...otherEntityApiImports(entity, service),
    ...complexTypeImports(entity.properties),
    ...enumTypeImports(entity.properties),
    odataImport(
      [
        'CustomField',
        'defaultDeSerializers',
        'DefaultDeSerializers',
        'DeSerializers',
        'AllFields',
        'entityBuilder',
        'EntityBuilderType',
        'EntityApi',
        'FieldBuilder',
        ...propertyTypeImportNames(entity.properties),
        ...propertyFieldTypeImportNames(entity.properties),
        ...navPropertyFieldTypeImportNames(
          entity.navigationProperties,
          service.oDataVersion
        )
      ],
      service.oDataVersion
    )
  ];
}

function otherEntityApiImports(
  entity: VdmEntity,
  service: VdmServiceMetadata
): Import[] {
  return Array.from(new Set(entity.navigationProperties.map(n => n.to)))
    .map(to => {
      const matchedEntity = service.entities.find(e => e.entitySetName === to);
      if (!matchedEntity) {
        throw Error(
          `Failed to find the entity from the service: ${JSON.stringify(
            service
          )} for entity ${entity}`
        );
      }
      return matchedEntity.className;
    })
    .filter(name => name !== entity.className)
    .flatMap(name => otherEntityImports(name));
}

function otherEntityImports(name: string): Import[] {
  return [
    {
      names: [`${name}Api`],
      moduleIdentifier: `./${name}Api`
    }
  ];
}
