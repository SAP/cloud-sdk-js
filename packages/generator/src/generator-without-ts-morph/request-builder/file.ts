import { serializeImports } from '@sap-cloud-sdk/generator-common/internal';
import { unixEOL } from '@sap-cloud-sdk/util';
import { requestBuilderClass } from './class';
import { requestBuilderImportDeclarations } from './imports';
import type { VdmEntity } from '../../vdm-types';
import type { ODataVersion } from '@sap-cloud-sdk/util';
import type { CreateFileOptions } from '@sap-cloud-sdk/generator-common/internal';

/**
 * @internal
 */
export function requestBuilderSourceFile(
  entity: VdmEntity,
  oDataVersion: ODataVersion,
  options?: CreateFileOptions
): string {
  const imports = serializeImports(
    requestBuilderImportDeclarations(entity, oDataVersion, options)
  );
  const content = requestBuilderClass(entity);
  return [imports, content].join(unixEOL);
}
