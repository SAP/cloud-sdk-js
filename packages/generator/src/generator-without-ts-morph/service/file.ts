import { codeBlock } from '@sap-cloud-sdk/util';
import {
  serializeImports,
  Import
} from '@sap-cloud-sdk/generator-common/internal';
import { hasEntities } from '../../generator-utils';
import { VdmServiceMetadata } from '../../vdm-types';
import { serviceBuilder, serviceClass } from './class';

/**
 * @internal
 */
export function serviceFile(service: VdmServiceMetadata): string {
  return codeBlock`${serializeImports(imports(service))}
  
${serviceBuilder(service.className, service.oDataVersion)} 
${serviceClass(service)}`;
}

function getImports(
  actionFunctionImports:
    | { name: string; parametersTypeName: string }[]
    | undefined,
  moduleIdentifier: 'function-imports' | 'action-imports'
): Import[] {
  if (actionFunctionImports === undefined) {
    return [];
  }
  const names = actionFunctionImports.map(
    actionFunctionImport => actionFunctionImport.name
  );
  const parameterNames = actionFunctionImports.map(
    actionFunctionImport => actionFunctionImport.parametersTypeName
  );
  return [
    {
      names: [...names, ...parameterNames],
      moduleIdentifier: `./${moduleIdentifier}`
    }
  ];
}

/**
 * @internal
 */
export function imports(service: VdmServiceMetadata): Import[] {
  const functionImports = getImports(
    service.functionImports,
    'function-imports'
  );

  const actionImports = getImports(service.actionImports, 'action-imports');
  const serviceImports = [
    ...service.entities.map(entity => ({
      names: [`${entity.className}Api`],
      moduleIdentifier: `./${entity.className}Api`
    })),
    ...functionImports,
    ...actionImports,
    {
      names: ['BigNumber'],
      moduleIdentifier: 'bignumber.js'
    },
    {
      names:
        service.oDataVersion === 'v2' ? ['Moment'] : ['Moment', 'Duration'],
      moduleIdentifier: 'moment'
    },
    {
      names: [
        'defaultDeSerializers',
        'DeSerializers',
        'DefaultDeSerializers',
        'mergeDefaultDeSerializersWith',
        'Time'
      ],
      moduleIdentifier: `@sap-cloud-sdk/odata-${service.oDataVersion}`
    }
  ];

  if (hasEntities(service)) {
    serviceImports.push({
      names: ['batch', 'changeset'],
      moduleIdentifier: './BatchRequest'
    });
  }

  return serviceImports;
}
