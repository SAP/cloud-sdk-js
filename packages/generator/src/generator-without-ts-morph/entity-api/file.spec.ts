import {
  breakfastEntity,
  foodService
} from '../../../test/test-util/data-model';
import { entityApiFile } from './file';

describe('entity api file', () => {
  it('gets proper file', () => {
    expect(entityApiFile(breakfastEntity, foodService)).toMatchSnapshot();
  });
});
