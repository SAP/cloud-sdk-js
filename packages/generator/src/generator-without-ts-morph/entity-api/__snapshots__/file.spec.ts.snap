// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity api file gets proper file 1`] = `
"import { Breakfast } from './Breakfast';
import { BreakfastRequestBuilder } from './BreakfastRequestBuilder';
import { BrunchApi } from './BrunchApi';
import { CustomField, defaultDeSerializers, DefaultDeSerializers, DeSerializers, AllFields, entityBuilder, EntityBuilderType, EntityApi, FieldBuilder, Time, OrderableEdmTypeField, OneToOneLink } from '@sap-cloud-sdk/odata-v2';
export class BreakfastApi<DeSerializersT extends DeSerializers = DefaultDeSerializers> implements 
    EntityApi<
      Breakfast<
        DeSerializersT
      >, 
      DeSerializersT
    > {
  public deSerializers: DeSerializersT;

  constructor(
    deSerializers: DeSerializersT = defaultDeSerializers as any) {
    this.deSerializers = deSerializers;
  }

  private navigationPropertyFields!: {
      /**
       * Static representation of the one-to-one navigation property {@link toBrunch} for query construction.
       * Use to reference this property in query operations such as 'select' in the fluent request API.
       */
      TO_BRUNCH: OneToOneLink<
            Breakfast<DeSerializersT>,
            DeSerializersT,
            BrunchApi<DeSerializersT>
          >
    };

  _addNavigationProperties(
      linkedApis: [
        BrunchApi<DeSerializersT>
      ]): this {
        this.navigationPropertyFields = {
          TO_BRUNCH: new OneToOneLink(
              'TO_BRUNCH',
              this,
              linkedApis[0]
            )
        };
        return this;
      }
  
  entityConstructor = Breakfast;
  
  requestBuilder(): BreakfastRequestBuilder<
    DeSerializersT
  > {
    return new BreakfastRequestBuilder<DeSerializersT>(this);
  }
  
  entityBuilder(): EntityBuilderType<
    Breakfast<
      DeSerializersT
    >,
    DeSerializersT
  > {
    return entityBuilder<Breakfast<DeSerializersT>, DeSerializersT>(this);
  }

  customField<NullableT extends boolean = false>(
    fieldName: string,
    isNullable: NullableT = false as NullableT
  ): CustomField<
  Breakfast<
      DeSerializersT>,
    DeSerializersT,
    NullableT
  > {
    return new CustomField(
      fieldName,
      this.entityConstructor,
      this.deSerializers,
      isNullable
    ) as any;
  }

  private _fieldBuilder?: FieldBuilder<typeof Breakfast, DeSerializersT>;
  get fieldBuilder() {
    if(!this._fieldBuilder){
      this._fieldBuilder = new FieldBuilder(Breakfast, this.deSerializers);
    }
    return this._fieldBuilder;
  }

  private _schema?: {
    ENTITY_NAME: OrderableEdmTypeField<Breakfast<DeSerializers>, DeSerializersT, 'Edm.String', false, true>,
NUMBER_OF_EGGS: OrderableEdmTypeField<Breakfast<DeSerializers>, DeSerializersT, 'Edm.Decimal', true, true>,
BREAKFAST_TIME: OrderableEdmTypeField<Breakfast<DeSerializers>, DeSerializersT, 'Edm.DateTime', false, true>,
/**
 * Static representation of the one-to-one navigation property {@link toBrunch} for query construction.
 * Use to reference this property in query operations such as 'select' in the fluent request API.
 */
TO_BRUNCH: OneToOneLink<
      Breakfast<DeSerializersT>,
      DeSerializersT,
      BrunchApi<DeSerializersT>
    >,
ALL_FIELDS: AllFields<Breakfast<DeSerializers>>
  };

  get schema() {
    if (!this._schema) {
      const fieldBuilder = this.fieldBuilder;
      this._schema = { 
    /**
 * Static representation of the {@link entityName} property for query construction.
 * Use to reference this property in query operations such as 'select' in the fluent request API.
 */
ENTITY_NAME: fieldBuilder.buildEdmTypeField('EntityName', 'Edm.String', false),
/**
 * Static representation of the {@link numberOfEggs} property for query construction.
 * Use to reference this property in query operations such as 'select' in the fluent request API.
 */
NUMBER_OF_EGGS: fieldBuilder.buildEdmTypeField('A_noEggs', 'Edm.Decimal', true),
/**
 * Static representation of the {@link breakfastTime} property for query construction.
 * Use to reference this property in query operations such as 'select' in the fluent request API.
 */
BREAKFAST_TIME: fieldBuilder.buildEdmTypeField('BreakfastTime', 'Edm.DateTime', false),
...this.navigationPropertyFields,
/**
 * 
 * All fields selector.
 */
ALL_FIELDS: new AllFields('*', Breakfast) 
  };
    }
  
    return this._schema;
  }
}"
`;
